### DISCUSSION TITLE
Tuya TS0601_dimmer_5

### DISCUSSION THREAD

**Discussion by JohnnyBreslin** [Koenkk/zigbee2mqtt#22027] [Devices]
Hi,

I have added a MOES single gang dimming light switch. It interviews and it is discovered as a TS0601_dimmer_5. 

The device responds to state "on" and "off" but does not respond to a state query("") or "toggle".

The Tuya device found is _TZE204_dcnsggvz. The device page on the zigbee2mqtt device page is clearly wrong as it shows a 2 gang wifi device but the commands look appropriate and include query state and toggle state in the examples.

I have updated zigbee2mqtt to latestrelease  and rebuilt. 

The error responses to {"state": "toggle"} and {"state": ""} are 
- Publish 'set' 'state' to 'hall' failed: 'Error: Value: 'TOGGLE' not found in: [ON, OFF]'
- Publish 'set' 'state' to 'hall' failed: 'Error: Value: '' not found in: [ON, OFF]':

I have attached the db entry below.

{"id":20, "type":"Router", "ieeeAddr":"0xf082c0fffe18cba8","nwkAddr":63497,"manufId":4098,"manufName":"_TZE204_dcnsggvz", "powerSource":"Mains (single phase)", "modelId":"TS0601","epList":[1],"endpoints":{"1":{"profId":260,"epId":1,"devId":81,"inClusterList":[0,4,5,61184],"outClusterList":[25,10],"clusters":{"genBasic":{"attributes":{"65503":"\u0000\u0000\u0000\u0000\u0005","65506":31,"65508":0,"65534":0,"modelId":"TS0601","manufacturerName":"_TZE204_dcnsggvz","stackVersion":0,"dateCode":"","zclVersion":3,"appVersion":70,"powerSource":1}}},"binds":[],"configuredReportings":[],"meta":{}}},"appVersion":70,"stackVersion":0,"hwVersion":1,"dateCode":"","zclVersion":3,"interviewCompleted":true,"meta":{"configured":-708457359},"lastSeen":1711997150212}

I am running node v18.11.0 on a Raspberrypi 4 with 8GB and CC2531 usb radio.

Any help would be greatly appreciated.

Thanks
JohnB

**Comment by schauveau** [comment]
The `state` attribute of `TS0601_dimmer_5` is defined here 

https://github.com/Koenkk/zigbee-herdsman-converters/blob/7727e110580a38c913cbd30f8c00ada394b0ad54/src/devices/tuya.ts#L1665
https://github.com/Koenkk/zigbee-herdsman-converters/blob/7727e110580a38c913cbd30f8c00ada394b0ad54/src/devices/tuya.ts#L1670

Most tuya devices are using the standard on-off cluster but here this is done via a tuya attribute.
This is probably why toggle does not work. 

So you could try `previous` instead of `toggle` since it is listed in the expose but I do not believe that this is going to work because `tuya.valueConverter.onOff` does not seem to support `previous`  

https://github.com/Koenkk/zigbee-herdsman-converters/blob/7727e110580a38c913cbd30f8c00ada394b0ad54/src/lib/tuya.ts#L562

The converter on the next line is `powerOnBehavior: valueConverterBasic.lookup({'off': 0, 'on': 1, 'previous': 2}),` and it supports `previous` so if we assume that the value 2 corresponds to toggle (which may not be true) then you could try the following quick and dirty experiment:
  - open the file `/opt/zigbee2mqtt/node_modules/zigbee-herdsman-converters/devices/tuya.js` in a text editor
  - search for the  `TS0601_dimmer_5` converter.
  - replace `onOff` by `powerOnBehavior` in the tuyaDatapoint for `state`. 
  - restart Z2M

You should now be able to try `{"state": "previous"}` without an error and if the device toggles then congratulation,  you just proved that the value `2` does a toggle (it may also behave as "on", "off" or do nothing). 

If it does not work then the toggle must be done manually for example in Home Assistant or by replacing a the `onOff` converter by a custom one that invert the current value of the`state` attribute.

Do not forget to restore 'onOff' in the `state` tuyaDatapoint once you are done with the experiment.
